control-plane/api-gateway/controllers/gateway_class_config_controller.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/consul-k8s/control-plane/api/v1alpha1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	gatewayClassConfigFinalizer = "gateway-class-exists-finalizer.consul.hashicorp.com"
+)
+
+type Controller struct {
+	client.Client
+
+	Log logr.Logger
+}
+
+// Reconcile reads the state of an Endpoints object for a Kubernetes Service and reconciles Consul services which
+// correspond to the Kubernetes Service. These events are driven by changes to the Pods backing the Kube service.
+func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("Reconcile the Gateway Class Config Controller", "name", req.Name, "namespace", req.Namespace)
+
+	gcc := &v1alpha1.GatewayClassConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gcc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.Log.Info("gateway class not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "failed to get gateway class config")
+		return ctrl.Result{}, err
+	}
+
+	if !gcc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// We have a deletion, ensure we're not in use.
+		used, err := r.GatewayClassConfigInUse(ctx, gcc)
+		if err != nil {
+			r.Log.Error(err, "failed to check if the gateway class config is still in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			r.Log.Info("gateway class config still in use")
+			// Requeue as to not block the reconciliation loop.
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		// gcc is no longer in use.
+		if _, err := r.RemoveFinalizer(ctx, gcc, gatewayClassConfigFinalizer); err != nil {
+			r.Log.Error(err, "error removing gateway class config finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if _, err := r.EnsureFinalizer(ctx, gcc, gatewayClassConfigFinalizer); err != nil {
+		r.Log.Error(err, "error adding gateway class config finalizer")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Controller) EnsureFinalizer(ctx context.Context, object client.Object, finalizer string) (bool, error) {
```

Comment:
👍 on making this private. In addition, can we actually just pass the kubernetes client as an argument to the method and make it a standalone method rather than being bound to the controller struct? We'll need something like this helper on our other gateway controllers as well, so might as well make this reusable from the get-go.

Reply:
If you want to keep this as-is for now, I will add the EnsureFinalizer func and you can leverage it in a separate PR.

Reply:
I need it for GatewayClass Controller anyways